fs/copy.py

```diff
@@ -332,54 +363,128 @@ def copy_dir_if_newer(
-                ]
-                for dir_path, copy_info in src_state:
+                for dir_path in walker.files(_src_fs, _src_path):
                     copy_path = combine(_dst_path, frombase(_src_path, dir_path))
-                    if copy_info.is_dir:
-                        _dst_fs.makedir(copy_path, recreate=True)
-                    elif copy_info.is_file:
-                        # dst file is present, try to figure out if copy
-                        # is necessary
-                        try:
-                            src_modified = copy_info.modified
-                            dst_modified = dst_state[dir_path].modified
-                        except KeyError:
-                            do_copy = True
-                        else:
-                            do_copy = (
-                                src_modified is None
-                                or dst_modified is None
-                                or src_modified > dst_modified
-                            )
-
-                        if do_copy:
-                            copier.copy(_src_fs, dir_path, _dst_fs, copy_path)
-                            on_copy(_src_fs, dir_path, _dst_fs, copy_path)
+                    if _copy_is_necessary(
+                        _src_fs, dir_path, _dst_fs, copy_path, condition
+                    ):
+                        copier.copy(_src_fs, dir_path, _dst_fs, copy_path)
+                        on_copy(_src_fs, dir_path, _dst_fs, copy_path)
+
+
+def _copy_is_necessary(
+    src_fs,  # type: FS
+    src_path,  # type: Text
+    dst_fs,  # type: FS
+    dst_path,  # type: Text
+    condition,  # type: Text
+):
+    # type: (...) -> bool
+
+    if condition == "always":
+        return True
+
+    elif condition == "newer":
+        try:
+            namespace = ("details", "modified")
+            src_modified = src_fs.getinfo(src_path, namespace).modified
+            dst_modified = dst_fs.getinfo(dst_path, namespace).modified
+        except ResourceNotFound:
+            return True
+        else:
+            return (
+                src_modified is None
+                or dst_modified is None
+                or src_modified > dst_modified
+            )
+
+    elif condition == "older":
+        try:
+            namespace = ("details", "modified")
+            src_modified = src_fs.getinfo(src_path, namespace).modified
+            dst_modified = dst_fs.getinfo(dst_path, namespace).modified
+        except ResourceNotFound:
+            return True
+        else:
+            return (
+                src_modified is None
+                or dst_modified is None
+                or src_modified < dst_modified
+            )
+
+    elif condition == "exists":
+        return dst_fs.exists(dst_path)
+
+    elif condition == "not_exists":
+        return not dst_fs.exists(dst_path)
+
+    else:
+        raise ValueError(condition + "is not a valid copy condition.")
```

Comment:
Please use string formatting here, in case `condition` is not a string this will fail (plus, you are missing a whitespace :wink: )

Reply:
Done. Didn't know that Python 2 already supported that.